### ISSUE TITLE
INSERT statement does not apply types to bind values in VALUES clause when using valuesOfRows()

### ISSUE THREAD

**Issue by cmrgKoradir** [jOOQ/jOOQ#14694]
### Use case

I want to write a query like this:

```kotlin
ctx
    .insertInto(
        SUMMARY_FORMAT_ELEMENT,
        SUMMARY_FORMAT_ELEMENT.FORMAT_ID,
        SUMMARY_FORMAT_ELEMENT.POSITION,
        SUMMARY_FORMAT_ELEMENT.CONTENT_ID,
        SUMMARY_FORMAT_ELEMENT.REFERENCE_MOLECULE_CLASS_ID,
        SUMMARY_FORMAT_ELEMENT.REFERENCE_POSITION,
    )
    .valuesOfRows(elements.withIndex().map { (position, formatElement) ->
        when (formatElement) {
            is StaticFormatElement -> DSL.row(
                formatId,
                position,
                formatElement.content.id,
                null,
                null,
            )
            is VariablePositionFormatElement -> DSL.row(
                formatId,
                position,
                null,
                formatElement.referenceMoleculeClass.id,
                formatElement.referencePosition,
            )
        }
    })
    .execute()
```

but that runs into
```
PSQLException: ERROR: UNION types text and integer cannot be matched
```
because postgres wrongfully assumes the `null` values are of type `TEXT` where I want them to be `INTEGER`s.

### Possible solution

Cast the `null` values behind the scenes to the type expected by the insert column.

### Possible workarounds

I *can* cast the null values manually, but then I can no longer use `.valuesOfRows` because `DSL::castNull` returns a `Field` and row assumes raw values.

Which makes the query a bit more complicated and more importantly, forces me to insert a lot of `DSL::value` calls:

```kotlin
ctx
    .insertInto(
        SUMMARY_FORMAT_ELEMENT,
        SUMMARY_FORMAT_ELEMENT.FORMAT_ID,
        SUMMARY_FORMAT_ELEMENT.POSITION,
        SUMMARY_FORMAT_ELEMENT.CONTENT_ID,
        SUMMARY_FORMAT_ELEMENT.REFERENCE_MOLECULE_CLASS_ID,
        SUMMARY_FORMAT_ELEMENT.REFERENCE_POSITION,
    )
    .let { insertInto ->
        elements.withIndex().fold(insertInto) { query, (position, formatElement) ->
            when (formatElement) {
                is StaticFormatElement -> query.values(
                    DSL.value(formatId),
                    DSL.value(position),
                    DSL.value(formatElement.content.id),
                    DSL.castNull(SUMMARY_FORMAT_ELEMENT.REFERENCE_MOLECULE_CLASS_ID),
                    DSL.castNull(SUMMARY_FORMAT_ELEMENT.REFERENCE_POSITION),
                )
                is VariablePositionFormatElement -> query.values(
                    DSL.value(formatId),
                    DSL.value(position),
                    DSL.castNull(SUMMARY_FORMAT_ELEMENT.CONTENT_ID),
                    DSL.value(formatElement.referenceMoleculeClass.id),
                    DSL.value(formatElement.referencePosition),
                )
            }
        }
    }
    .execute()
```

which is annoying, because I already fed jOOQ all the information it needs to be able to cast correctly in the `.insertInto` step.

### jOOQ Version

Pro 3.17.6

### Database product and version

PostgreSQL 13.5 on x86_64-pc-linux-musl, compiled by gcc (Alpine 10.3.1_git20211027) 10.3.1 20211027, 64-bit

**Comment by lukaseder** [comment]
Thanks for your report. Curious, that should work already. There are some harder to fix cases, like https://github.com/jOOQ/jOOQ/issues/11015.

But this one should already work, I think it does when you pass records instead (to `valuesOfRecords()`)

**Comment by lukaseder** [comment]
I'm curious about how this really happens. Given:

```sql
create table t (i int);
```

All of these work:

```sql
insert into t (i)
values (null),(null),(1);

insert into t (i)
select null as a
union all
select 1 as a;
```

But this one fails:

```sql
insert into t (i)
select null as a
union all
select null as a
union all
select 1 as a;
```

While I still think that last query should work as well, did you perhaps omit a detail of how you construct your query?

**Comment by lukaseder** [comment]
E.g. before PostgreSQL 9.5, `INSERT .. ON CONFLICT DO NOTHING` was emulated with `INSERT .. SELECT .. WHERE NOT EXISTS ..`, but you're not using such an old PG version...

**Comment by lukaseder** [comment]
One case where I can reproduce this is with with client side stored computed columns (#9879), which must be emulated with `INSERT .. SELECT` on all dialects. Are you using that feature on this particular table?

**Comment by cmrgKoradir** [comment]
We are using audit columns (which are not listed in the `.insertInto` step), yes.

Or were at the time I opened the issue, at least.

I was able to slim down that table after migration to jOOQ and can confirm that getting rid of the audit columns solved the issue for this particular table.

Obviously, I will not be able to just remove them on other tables.

Will try with `valuesOfRecords` next time I run into this, but I do think long-term, `valuesOfRows` still should be able to handle this.

Thank you.

**Comment by lukaseder** [comment]
Thanks for the confirmation. Audit columns are built on top of client side stored computed columns, so indeed, this would explain it.

(Off topic, I haven't heard much feedback yet regarding this audit column feature. How well does it work for you? Do you have any feedback / feature requests to share? Would be very interested (in a new issue).)

> Will try with `valuesOfRecords` next time I run into this, but I do think long-term, `valuesOfRows` still should be able to handle this.

Sure. I'm fixing it right now. The fix will be available in the next set of patch releases, probably next week.

**Comment by cmrgKoradir** [comment]
> (Off topic, I haven't heard much feedback yet regarding this audit column feature. How well does it work for you? Do you have any feedback / feature requests to share? Would be very interested (in a new issue).

Well, I am actually writing a bug report about them, right now. :sweat_smile: 

Works great for `auditInsertUser` and `auditUpdateUser`, but we just found out that jOOQ ignores our `auditInsertTimestamp` and `auditUpdateTimestamp` configurations ... which we didn't notice until now because we rarely display the modification date, the creation timestamp doesn't change and we have default values on both of the timestamp columns. :man_facepalming:

**Comment by lukaseder** [comment]
Oh no, that's not good. "Looking forward" to your bug report, then!

**Comment by lukaseder** [comment]
Related to this issue here, while it would be possible to focus only on the `null` case, when `row(null, x)` is used in a `valuesOfRows()` clause, the real problem might as well be that we don't apply the usual coercion of the known `INSERT` columns type to any row contents. For example, while this test passes:

```java
public void testConverterExpressions_INSERT() throws Exception {
    bindAndInline(ctx -> {
        try {
            Field<String> l = TAuthor_LAST_NAME().convert(String.class,
                t -> t == null ? null : "X:" + t,
                u -> u == null ? null : u.replace("X:", "")
            );

            assertEquals(1,
                ctx.insertInto(TAuthor())
                   .columns(TAuthor_ID(), l)
                   .values(3, "X:A")
                   .execute());

            assertEquals("X:A", ctx.select(l).from(TAuthor()).where(TAuthor_ID().eq(3)).fetchOne(r -> r.value1()));
            assertEquals("A", ctx.select(TAuthor_LAST_NAME()).from(TAuthor()).where(TAuthor_ID().eq(3)).fetchOne(r -> r.value1()));
        }
        finally {
            resetAuthors();
        }
    });
}
```

This test fails:

```java
public void testConverterExpressions_INSERT_VALUES_OF_ROWS() throws Exception {
    bindAndInline(ctx -> {
        try {
            Field<String> l = TAuthor_LAST_NAME().convert(String.class,
                t -> t == null ? null : "X:" + t,
                u -> u == null ? null : u.replace("X:", "")
            );

            assertEquals(1,
                ctx.insertInto(TAuthor())
                   .columns(TAuthor_ID(), l)
                   .valuesOfRows(asList(row(3, "X:A")))
                   .execute());

            assertEquals("X:A", ctx.select(l).from(TAuthor()).where(TAuthor_ID().eq(3)).fetchOne(r -> r.value1()));
            assertEquals("A", ctx.select(TAuthor_LAST_NAME()).from(TAuthor()).where(TAuthor_ID().eq(3)).fetchOne(r -> r.value1()));
        }
        finally {
            resetAuthors();
        }
    });
}
```

The rule here should be that if any `T` value is passed to a `row()` constructor (as opposed to passing a `Field<T>`), we should remember that the implicitly created `Param<T>` didn't have the actual type information available yet. Once it becomes available (e.g. because the `row()` is used in an `INSERT`), then we should apply the known type to the `T` value, instead of the default type.

In my opinion, this should automatically solve the `null` problem as well, at least for this case.

**Comment by lukaseder** [comment]
It's again different with `valuesOfRecords()`, because there, the assumption is that the record already has the right `Field<T> : T` key/value association. But with `valuesOfRows()`, while the Java compiler infers types correctly, the runtime doesn't apply them correctly.

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.18.0, 3.17.9 (https://github.com/jOOQ/jOOQ/issues/14705), and 3.16.15 (https://github.com/jOOQ/jOOQ/issues/14706)

**Comment by lukaseder** [comment]
This fix probably caused the following regression:

- https://github.com/jOOQ/jOOQ/issues/17802
